avro4s-core/src/main/scala/com/sksamuel/avro4s/TypeUnions.scala

```diff
@@ -0,0 +1,124 @@
+      }
+    case _ => sys.error(s"Unsupported type $value in type union decoder")
+  }
+}
+
+object TypeUnions {
+
+  def encoder[T](ctx: SealedTrait[Encoder, T],
+                 env: DefinitionEnvironment[Encoder],
+                 update: SchemaUpdate): Encoder[T] = {
+    // cannot extend the recursive environment with an initial type union encoder with empty union schema, as Avro Schema
+    // doesn't support this. So we use the original recursive environment to build subtypes, meaning that in case of a
+    // recursive schema, two identical type union encoders may be created instead of one.
+    val subtypeEncoders = enrichedSubtypes(ctx, update).map { case (st, u) => new UnionEncoder[T](st)(env, u) }
+    val schemaFor = buildSchema[T](update, subtypeEncoders.map(_.schema))
+    val encoderBySubtype = subtypeEncoders.map(e => e.subtype -> e).toMap
+    new TypeUnionEncoder[T](ctx, schemaFor, encoderBySubtype)
+  }
+
+  def decoder[T](ctx: SealedTrait[Decoder, T],
+                 env: DefinitionEnvironment[Decoder],
+                 update: SchemaUpdate): Decoder[T] = {
+    // cannot extend the recursive environment with an initial type union decoder with empty union schema, as Avro Schema
+    // doesn't support this. So we use the original recursive environment to build subtypes, meaning that in case of a
+    // recursive schema, two identical type union decoders may be created instead of one.
+    val subtypeDecoders = enrichedSubtypes(ctx, update).map { case (st, u) => new UnionDecoder[T](st)(env, u) }
+    val schemaFor = buildSchema[T](update, subtypeDecoders.map(_.schema))
+    val decoderByName = subtypeDecoders.map(decoder => decoder.fullName -> decoder).toMap
+    new TypeUnionDecoder[T](ctx, schemaFor, decoderByName)
+  }
+
+  def schema[T](ctx: SealedTrait[SchemaFor, T],
+                env: DefinitionEnvironment[SchemaFor],
+                update: SchemaUpdate): SchemaFor[T] = {
+    val subtypeSchema: ((Subtype[SchemaFor, T], SchemaUpdate)) => Schema = {
+      case (st, NamespaceUpdate(ns))     => SchemaHelper.overrideNamespace(st.typeclass.resolveSchemaFor(env, update).schema, ns)
+      case (_, FullSchemaUpdate(schemaFor)) => schemaFor.schema
+      case (st, NoUpdate)          => st.typeclass.resolveSchemaFor(env, update).schema
+    }
+
+    val subtypeSchemas = enrichedSubtypes(ctx, update).map(subtypeSchema)
+    SchemaFor[T](SchemaHelper.createSafeUnion(subtypeSchemas: _*), DefaultFieldMapper)
+  }
+
+  private def enrichedSubtypes[Typeclass[_], T](ctx: SealedTrait[Typeclass, T],
+                                                update: SchemaUpdate): Seq[(Subtype[Typeclass, T], SchemaUpdate)] = {
+    val enrichedUpdate = update match {
+      case NoUpdate =>
+        // in case of namespace annotations, pass the namespace update down to all subtypes
+        val ns = new AnnotationExtractors(ctx.annotations).namespace
+        ns.fold[SchemaUpdate](NoUpdate)(NamespaceUpdate)
+      case _ => update
```

Comment:
Would the annotated namespace be missed here since if we have a namespace/full update, we don't inspect the trait? I guess this means we're just assuming you annotate the trait itself, which I think is fine.

Reply:
The code gives priority to explicit schema customizations (via `withSchema`) over annotations on the trait. If the trait isn't annotated but some of the elements are, these namespace annotations will be handled in `Records`.